File: PublicSequenceMemoizer/src/main/java/edu/columbia/stat/wood/pub/sequencememoizer/IntSequenceMemoizerInterface.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */

package edu.columbia.stat.wood.pub.sequencememoizer;

import edu.columbia.stat.wood.pub.sequencememoizer.util.IntDiscreteDistribution;

/**
 * Sequence memoizer interface for int types.
 * @author nicholasbartlett
 */
public interface IntSequenceMemoizerInterface {

    /**
     * Indicates to the model that you wish to start a new sequence.  The first
     * time continueSequence is used after this the assumption is that the type
     * is being seen with empty context.
     */
    public void newSequence();

    /**
     * Continue the current sequence by incorporating the new observation.
     * @param observation
     * @return log predictive probability of observation prior to incorporating the observation into the model
     */
    public double continueSequence(int observation);

    /**
     * Continue the sequence with the sequence of observations supplied.  This is
     * the same as looping over the method call continueSequence for each observation.
     * @param observations
     * @return the sum of the log predictive probabilities of each observation prior to incorporating it into the model
     */
    public double continueSequence(int[] observations);

    /**
     * Generate some number of samples given a specified context.
     * @param context context, arranged most distant to the left, most recent to the right
     * @param numSamples number of samples desired
     * @return samples from the predictive distribution in the specified context
     */
    public int[] generate(int[] context, int numSamples);

    /**
     * Generate a sequence from the predictive model, starting with the specified context.
     * @param context starting context
     * @param sequenceLength length of sequence
     * @return sampled sequence of specified length
     */
    public int[] generateSequence(int[] context, int sequenceLength);

    /**
     * Get the predictive distribution in a specified context.
     * @param context
     * @return discrete distribution over integer types
     */
    public IntDiscreteDistribution predictiveDistribution(int[] context);

    /**
     * Get the predictive probability of a given type in a given context.
     * @param context
     * @param type
     * @return probability of type in the specified context
     */
    public double predictiveProbability(int[] context, int type);

    /**
     * Gets the log probability of a sequence given the current state of the model.
     * @param context context prior to the start of the sequence to score
     * @param sequence
     * @return log probability of sequence in specified context.
     */
    public double sequenceProbability(int[] context , int[] sequence);

    /**
     * Sample the parameters in the model.  The seating arrangements are sampled
     * using Gibbs sampling while the discount parameters use a Gibbs-Metropolis step.
     * @param numSweeps number of sweeps for the Gibbs sampler
     * @return joint log probability of model and data
     */
    public double sample(int numSweeps);

    /**
     * Sample only the seating arrangements.
     * @param numSweeps
     */
    public void sampleSeatingArrangements(int numSweeps);

    /**
     * Sample only the discounts.
     * @param numSweeps
     * @return joint log probability of model and data
     */
    public double sampleDiscounts(int numSweeps);

    /**
     * Get the joint log probability of model and data.
     * @return joint log probability of model and data
     */
    public double score();

    /**
     * Get the parameters of this model.
     * @return parameters of this model
     */
    public IntSequenceMemoizerParameters getParameters();

}

File: R_SequenceMemoizer/R/IntSequenceMemoizerInterface.R
#' Class IntSequenceMemoizerInterface
#'
#' Implementation of a sequence memoizer which takes int types. 
#'
#' @param sm a edu.columbia.stat.wood.pub.sequencememoizer.IntSequenceMemoizer object
#' @param ismp a edu.columbia.stat.wood.pub.sequencememoizer.IntSequenceMemoizerParameters object
setRefClass ("IntSequenceMemoizerInterface",
	fields = list (sm = "jobjRef", parameters = "IntSequenceMemoizerParameters"),
	methods = list (
		continueSequence = function (observations) {
			observations <- .as.integer (observations)
			return (.jcall (obj=sm, returnSig="D", method="continueSequence", .jarray (observations)))
		},
		generate = function (context, numSamples) {
			context <- .as.integer(context)
			numSamples <- .as.integer(numSamples)
			return (.jcall (obj=sm, returnSig="[I", method="generate", .jarray (context), numSamples))
		},
		generateSequence = function (context, sequenceLength) {
			context <- .as.integer(context)
			sequenceLength <- .as.integer(sequenceLength)
			return (.jcall (obj=sm, returnSig="[I", method="generateSequence", .jarray (context), sequenceLength))
		}, 
		getParameters = function () {
			return (parameters)
		},
		newSequence = function () {
			return (.jcall (obj=sm, returnSig="V", method="newSequence"))
		},
		## For some reason, rJava can not resolve IntDiscreteDistribution
		predictiveDistribution = function (context) {
			context <- .as.integer (context)
			#return (.jcall (obj=sm, returnSig="Ledu/columbia/stat/wood/pub/util/IntDiscreteDistribution", method="predictiveDistribution", .jarray (context)))
			return (sm$predictiveDistribution (.jarray (context)))
		},
		predictiveProbability = function (context, type) {
			context <- .as.integer(context)
			type <- .as.integer (type)
			return (.jcall (obj=sm, returnSig="D", method="predictiveProbability", .jarray (context), type))
		},
		sample = function (numSweeps) {
			numSweeps <- .as.integer (numSweeps)
			return (.jcall (obj=sm, returnSig="D", method="sample", numSweeps))
		},
		sampleDiscounts = function (numSweeps) {
			numSweeps <- .as.integer (numSweeps)
			return (.jcall (obj=sm, returnSig="D", method="sampleDiscounts", numSweeps))
		},
		sampleSeatingArrangements = function (numSweeps) {
			numSweeps <- .as.integer (numSweeps)
			return (.jcall (obj=sm, returnSig="V", method="sampleSeatingArrangements", numSweeps))
		},
		score = function () {
			return (.jcall (obj=sm, returnSig="D", method="score"))
		},
		sequenceProbability = function (context, sequence) {
			context <- .as.integer(context)
			sequence <- .as.integer (sequence)
			return (.jcall (obj=sm, returnSig="D", method="sequenceProbability", .jarray (context), .jarray (sequence)))
		}
	)
)